src/cockpit/channels/metrics.py

```diff
@@ -15,23 +15,148 @@
+                return
+
+            # Keep track of what samplers are requested
+            self.samplers |= sampler.value
+
+            desc = MetricDescription(name, unit, semantics, [], sampler.value)
+            self.metrics.append(MetricInfo(desc, derive, [], None))
+
+        self.ready()
+
+    async def run(self, options):
+        self.parse_options(options)
+
+        while True:
+            data = []
+            metrics = []
+            samples = Samples()
+            # everything is in milliseconds
+            timestamp = int(time.time() * 1000)
+            self.next_timestamp = timestamp
+
+            if Sampler.CPU_SAMPLER:
+                cockpit_cpu_samples(samples)
+            if Sampler.MEMORY_SAMPLER:
+                cockpit_memory_samples(samples)
+
+            for metricinfo in self.metrics:
+                name = metricinfo.desc.name
+                old_value = metricinfo.value
+                value = samples[name]
+
+                # TODO: constify 'rate'
+                if metricinfo.derive == "rate": 
+                    if old_value:
+                        data.append((value - old_value) / (self.next_timestamp - self.last_timestamp) * 1000)
+                    else:
+                        data.append(False)
+                else:
+                    data.append(value)
+
+                metrics.append(MetricInfo(desc=metricinfo.desc, derive=metricinfo.derive, instances=[], value=value))
+
+            self.metrics = metrics
+
+            if self.need_meta:
+                metrics = []
+                for metricinfo in self.metrics:
+                    metrics.append({
+                        'name': metricinfo.desc.name,
+                        'derive': metricinfo.derive,
+                        'units': metricinfo.desc.units,
+                        'semantics': metricinfo.desc.semantics
+                    })
+
+                meta = {
+                    'timestamp': timestamp,
+                    'interval': self.interval,
+                    'source': 'internal',
+                    'metrics': metrics
+                }
+                self.send_message(meta=meta)
+                self.need_meta = False
+
+            self.last_timestamp = self.next_timestamp
+            self.send_data(json.dumps([data]).encode())
+            logger.debug("sending interalmetrics data: %s", data)
 
-    def do_open(self, options):
-        assert options['source'] == 'internal'
-        assert options['interval'] == 3000
-        assert 'omit-instances' not in options
-        assert options['metrics'] == [
-            {"name": "cpu.basic.user", "derive": "rate"},
-            {"name": "cpu.basic.system", "derive": "rate"},
-            {"name": "cpu.basic.nice", "derive": "rate"},
-            {"name": "memory.used"},
-        ]
+            try:
+                ret = await asyncio.wait_for(self.read(), self.interval / 1000)
```

Comment:
@allisonkarlitskaya this never returns anything when the channel is closed so we keep sending metrics at the moment. You said something about `self.read` returning EOF but this doesn't seem to happen here.

Reply:
https://github.com/cockpit-project/cockpit/blob/main/src/cockpit/channel.py#L187

```python
    def do_done(self):
        self.receive_queue.put_nowait(b'')
```

the EOF gets written only if the other side says `done`.  We could probably modify the asyncchannel class to also send it on close.